### PR TITLE
nixos/duckdns: init module

### PR DESCRIPTION
**Pull request by notthebee**
## Description of changes

Added a Systemd service for updating DuckDNS domains

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/misc/duckdns.nix
Comment (bjornfor) [inline nixos/modules/services/misc/duckdns.nix:63]: Echoing the query string like this exposes the secret to all (unprivileged) processes on the system -- can we avoid that?

File: nixos/modules/misc/ids.nix
Comment (Scrumplex) [inline nixos/modules/misc/ids.nix:359]: Do we need this? I don't think systemd's `DynamicUser` will use these anyway

File: nixos/modules/misc/ids.nix
Comment (Scrumplex) [inline nixos/modules/misc/ids.nix:670]: same here ^^

File: nixos/modules/services/misc/duckdns.nix
Comment (Scrumplex) [inline nixos/modules/services/misc/duckdns.nix:7]: nit
```suggestion
  options.services.duckdns = {
```

File: nixos/modules/services/misc/duckdns.nix
Comment (Scrumplex) [inline nixos/modules/services/misc/duckdns.nix:1]: Maybe add yourself as a maintainer for this module? There is some information about this here: https://github.com/NixOS/nixpkgs/tree/master/maintainers (just add yourself to maintainer-list.nix and add `meta.maintainers = with lib.maintainers; [ <name> ];` to this module file

File: nixos/modules/services/misc/duckdns.nix
Comment (Scrumplex) [inline nixos/modules/services/misc/duckdns.nix:12]: This supports both Nix paths and strings

```suggestion
        type = lib.types.path;
```

File: nixos/modules/services/misc/duckdns.nix
Comment (Scrumplex) [inline nixos/modules/services/misc/duckdns.nix:13]: `lib.mdDoc` is a no-op now (it returns the value as-is), so it can be omitted

File: nixos/modules/services/misc/duckdns.nix
Comment (Scrumplex) [inline nixos/modules/services/misc/duckdns.nix:31]:
```suggestion
        type = lib.types.nullOr lib.types.path;
```

File: nixos/modules/services/misc/duckdns.nix
Comment (Scrumplex) [inline nixos/modules/services/misc/duckdns.nix:42]: I think this can use an assertion to make sure either `domains` or `domainsFile` are set
```suggestion
  config = lib.mkIf cfg.enable {
    assertions = [
      {
        assertion = (cfg.domains != null || cfg.domainsFile != null);
        message = "services.duckdns.domains or services.duckdns.domainsFile has to be defined";
      }
    ];
```

**Review by bjornfor — CHANGES_REQUESTED** [review]
Echoing the query string like this exposes the secret to all (unprivileged) processes on the system -- can we avoid that?

If not, please add a big warning in the option description.

**Comment by bjornfor** [comment]
You updated at the same time I commented 😀

However, the secret is still visible in process listings.

**Comment by bjornfor** [comment]
I think "curl -K ./secret_params.txt ..." where secret_params.txt contains lines of "-d key=value" is the way to go. "replace-secret" in nixpkgs might help with securely preparing the file. (Untested.)

**Comment by notthebee** [comment]
Used the -K parameter and some shell redirection to hide the secrets in the process list

**Comment by bjornfor** [comment]
> Used the -K parameter and some shell redirection to hide the secrets in the process list

Thanks!

Please rebase / squash the commits and I can take another look. (I think those 11 commits are just fixups to the one "init" commit?)

**Comment by notthebee** [comment]
> > Used the -K parameter and some shell redirection to hide the secrets in the process list
> 
> Thanks!
> 
> Please rebase / squash the commits and I can take another look. (I think those 11 commits are just fixups to the one "init" commit?)

Done :)

File: nixos/modules/services/misc/duckdns.nix
Comment (bjornfor) [inline nixos/modules/services/misc/duckdns.nix:65]: tokenFile defaults to null and there is no assert; I think that will result in a poor error message here.

File: nixos/modules/services/misc/duckdns.nix
Comment (bjornfor) [inline nixos/modules/services/misc/duckdns.nix:49]: I think you should also assert that not both are set, because I think the implementation will then only use one of them and discard the other.

File: nixos/modules/services/misc/duckdns.nix
Comment (bjornfor) [inline nixos/modules/services/misc/duckdns.nix:79]: That line looked scary (as in leaking secrets on the command line), but I looked it up (man bash), and tested it, and that string is only passed to stdin for the curl process.

File: maintainers/maintainer-list.nix
Comment (bjornfor) [inline maintainers/maintainer-list.nix:15632]: Please factor out this hunk into a "maintainers: add notthebee" commit, and order it before the "nixos/duckdns: init module" commit.

File: nixos/modules/services/misc/duckdns.nix
Comment (bjornfor) [inline nixos/modules/services/misc/duckdns.nix:12]: No need for parenthesis around the mkEnableOption argument string.

File: nixos/modules/services/misc/duckdns.nix
Comment (h7x4) [inline nixos/modules/services/misc/duckdns.nix:61]: Sorry, I know this is about the third time someone has commented on these assertions. Hopefully final time.

```suggestion
      {
        assertion = lib.xor (cfg.domains != null) (cfg.domainsFile != null);
        message = "Only one of services.duckdns.domains and services.duckdns.domainsFile can be defined";
      }
```

Comment (bjornfor) [reply nixos/modules/services/misc/duckdns.nix]: I think with the xor assert, the error message gets slightly confusing. I tried rewording it to `One of services.duckdns.domains and services.duckdns.domainsFile must be defined, but not both`, but I'm not happy with that either. (Think of the case when neither is set.) So I think I prefer separate assert/message. (But no strong feelings, just ignore if you want.)

Comment (h7x4) [reply nixos/modules/services/misc/duckdns.nix]: On one hand, it feels overkill having two assertions to cover for what is essentially one assertion (or at least it is in my head), but you do have a good point about the error message. I'll leave the decision to the PR author. Sorry about the additional noise and iteration

File: nixos/modules/services/misc/duckdns.nix
Comment (bjornfor) [inline nixos/modules/services/misc/duckdns.nix:28]: s/prefix/suffix/?

File: nixos/modules/services/misc/duckdns.nix
Comment (bjornfor) [inline nixos/modules/services/misc/duckdns.nix:92]: Although I don't know the full response from that curl command, I generally think it's better to check for success response ("OK") than the absence of error response ("KO").

Also, I'd feel better about the robustness of this code if errors from curl were detected, i.e. checking the exit code.

File: nixos/modules/services/misc/duckdns.nix
Comment (Scrumplex) [inline nixos/modules/services/misc/duckdns.nix:118]: nit:
```suggestion
        exec ${lib.getExe duckdns}
```

File: nixos/modules/services/misc/duckdns.nix
Comment (Scrumplex) [inline nixos/modules/services/misc/duckdns.nix:28]: I think we should return a non-zero exit code to make monitoring failures easier.
```suggestion
    else
        echo -e "Something went wrong, please check your settings\nThe response returned was:\n$DRESPONSE\n"
        exit 1
    fi
```

**Review by Scrumplex — APPROVED** [review]
Apart from these small nit picks, LGTM

**Review by Scrumplex — APPROVED** [review]

**Comment by notthebee** [comment]
@bjornfor @h7x4 Are there any other changes I should make before this can be merged?

**Review by bjornfor — APPROVED** [review]

**Review by Scrumplex — APPROVED** [review]

**Pull request merged**
